sklearn/cluster/_k_means.pyx

```diff
@@ -8,161 +8,257 @@
 #
 # License: BSD 3 clause
 
-from libc.math cimport sqrt
 import numpy as np
-import scipy.sparse as sp
 cimport numpy as np
 cimport cython
 from cython cimport floating
+from libc.math cimport sqrt
+
+from ..utils.extmath import row_norms
+
+
+np.import_array()
 
-from sklearn.utils.sparsefuncs_fast import assign_rows_csr
-from ..utils._cython_blas cimport _dot
 
 ctypedef np.float64_t DOUBLE
 ctypedef np.int32_t INT
 
 
-np.import_array()
+cdef floating _euclidean_dense_dense(floating* a,
+                                     floating* b,
+                                     int n_features,
+                                     bint squared) nogil:
+    """Euclidean distance between a dense and b dense"""
+    cdef:
+        int i
+        int n = n_features // 4
```

Comment:
That's for optimizing cache hit right? Maybe just add a comment to justify it?

Reply:
Yes, but it happens that it's not necessary if you compile with the `-Ofast` flag.

 I won't change it right now because ideally I'd like to use these from `pairwise_fast` to avoid duplicating code. They are not there yet but I have a PR for euclidean distances which should be merged before this one.